server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -106,31 +110,69 @@ public void getSegmentFiles(
                     indexShard.remoteStore().incRef();
                     final Directory storeDirectory = indexShard.store().directory();
+                    final ShardPath shardPath = indexShard.shardPath();
                     for (StoreFileMetadata fileMetadata : filesToFetch) {
                         String file = fileMetadata.name();
                         assert directoryFiles.contains(file) == false : "Local store already contains the file " + file;
-                        storeDirectory.copyFrom(remoteDirectory, file, file, IOContext.DEFAULT);
-                        downloadedSegments.add(fileMetadata);
+                        toDownloadSegments.add(fileMetadata);
                     }
-                    logger.trace("Downloaded segments from remote store {}", downloadedSegments);
+
+                    CountDownLatch countDownLatch = new CountDownLatch(1);
+                    LatchedActionListener<GetSegmentFilesResponse> completionListener = new LatchedActionListener<>(
+                        listener,
+                        countDownLatch
+                    );
+                    downloadSegments(storeDirectory, remoteDirectory, toDownloadSegments, shardPath, completionListener);
+                    countDownLatch.await();
+
+                    logger.trace("Downloaded segments from remote store {}", toDownloadSegments);
                 } finally {
                     indexShard.store().decRef();
                     indexShard.remoteStore().decRef();
                 }
             }
-            listener.onResponse(new GetSegmentFilesResponse(downloadedSegments));
         } catch (Exception e) {
             listener.onFailure(e);
         }
     }
 
+    private void downloadSegments(
+        Directory storeDirectory,
+        RemoteSegmentStoreDirectory remoteStoreDirectory,
+        List<StoreFileMetadata> toDownloadSegments,
+        ShardPath shardPath,
+        ActionListener<GetSegmentFilesResponse> completionListener
+    ) {
+
+        final AtomicInteger totalDownloadedSegments = new AtomicInteger(toDownloadSegments.size());
+        final Path indexPath = shardPath == null ? null : shardPath.resolveIndex();
+        final ActionListener<String> segmentsDownloadListener = new ActionListener<>() {
+            @Override
+            public void onResponse(String fileName) {
+                if (totalDownloadedSegments.decrementAndGet() == 0) {
+                    completionListener.onResponse(new GetSegmentFilesResponse(toDownloadSegments));
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error("Failed to download one of the segment files from the remote store", e);
+                completionListener.onFailure(e);
+            }
+        };
+
+        toDownloadSegments.forEach(
+            fileMetadata -> remoteStoreDirectory.copyTo(fileMetadata.name(), storeDirectory, indexPath, segmentsDownloadListener)
```

Comment:
Why do we need to do this sequentially? We can submit all files download in parallel. You can look at `uploadNewSegments` of RemoteStoreRefreshListener. We just need to send appropriate listeners for each file download callback and then group them with GroupedActionListener and just need a latch of size 1 if upstream flow is synchronous. In this case, it looks like we don't even need a 1 sized latch as we just need to send callback on the upstream listener.

Reply:
This is not sequential - it is parallel. I have updated the code to use a grouped action listener instead.